SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift

```diff
@@ -0,0 +1,250 @@
+        // Compare against new CDS service
+        let newCDSBatchOperation = CDSBatchOperation(recipientIdsToLookup: self.recipientIdsToLookup)
+        let cdsFeedbackOperation = CDSFeedbackOperation(legacyRegisteredRecipientIds: self.registeredRecipientIds)
+        cdsFeedbackOperation.addDependency(newCDSBatchOperation)
+
+        CDSFeedbackOperation.operationQueue.addOperations([newCDSBatchOperation, cdsFeedbackOperation], waitUntilFinished: false)
+    }
+
+    // MARK: Private Helpers
+
+    private func parse(response: Any?, phoneNumbersByHashes: [String: String]) throws -> Set<String> {
+
+        guard let responseDict = response as? [String: AnyObject] else {
+            let responseError: NSError = OWSErrorMakeUnableToProcessServerResponseError() as NSError
+            responseError.isRetryable = true
+
+            throw responseError
+        }
+
+        guard let contactDicts = responseDict["contacts"] as? [[String: AnyObject]] else {
+            let responseError: NSError = OWSErrorMakeUnableToProcessServerResponseError() as NSError
+            responseError.isRetryable = true
+
+            throw responseError
+        }
+
+        var registeredRecipientIds: Set<String> = Set()
+
+        for contactDict in contactDicts {
+            guard let hash = contactDict["token"] as? String, hash.count > 0 else {
+                owsFail("\(self.logTag) in \(#function) hash was unexpectedly nil")
+                continue
+            }
+
+            guard let recipientId = phoneNumbersByHashes[hash], recipientId.count > 0 else {
+                owsFail("\(self.logTag) in \(#function) recipientId was unexpectedly nil")
+                continue
+            }
+
+            guard recipientIdsToLookup.contains(recipientId) else {
+                owsFail("\(self.logTag) in \(#function) unexpected recipientId")
+                continue
+            }
+
+            registeredRecipientIds.insert(recipientId)
+        }
+
+        return registeredRecipientIds
+    }
+
+}
+
+class CDSBatchOperation: OWSOperation {
+
+    private let recipientIdsToLookup: [String]
+    var registeredRecipientIds: Set<String>
+
+    // MARK: Initializers
+
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+    }
+
+    // MARK: OWSOperationOverrides
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        Logger.debug("\(logTag) in \(#function) FAKING intersection (TODO)")
+        self.registeredRecipientIds = Set(self.recipientIdsToLookup)
+        self.reportSuccess()
+    }
+}
+
+class CDSFeedbackOperation: OWSOperation {
+
+    static let operationQueue = OperationQueue()
```

Comment:
Shouldn't we configure the concurrency level on this queue?

Reply:
Good catch! Fixed.